### PR TITLE
Novel: Flow-gradient-weighted surface loss (boundary layer physics prior)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
In CFD, the boundary layer determines surface pressure through the pressure gradient normal to the surface. Currently, ALL surface nodes are weighted equally in the loss. But physics tells us: nodes at high-curvature regions (leading edge, suction peak) carry most of the surface pressure information because that is where the pressure gradient is steepest. Weighting surface loss nodes by a proxy for local flow complexity — the VARIANCE of predictions among nearby nodes — should focus gradient signal on the hardest, most physically important regions.

**Why this differs from prior curvature-weighted loss (#920, #1470, #1506):** Those weighted by geometry (curvature). This weights by PREDICTION DIFFICULTY — measured at runtime from the model's own uncertainty. Nodes where the model's predictions vary most within a local neighborhood are the hardest to get right and carry the most information.

**Inspiration:** Anisotropic diffusion in image processing (Perona-Malik) smooths along edges but preserves edge gradients. We do the analogous thing for loss: upweight where the predicted field has high local variation (the "edges" of the flow field).

## Instructions
In the training loop, after computing `abs_err` (~line 710), add a prediction-variance weight for surface nodes:

```python
# Flow-gradient-weighted surface loss (replaces simple surf_per_sample for pressure)
if epoch >= 15:
    # Compute local prediction variance as hardness proxy
    surf_pred_p = pred[:, :, 2]  # [B, N] pressure predictions
    surf_pred_p_masked = surf_pred_p.masked_fill(~surf_mask, 0.0)
    # Use coarse pooling to get local-mean prediction, then compute deviation
    # Sort surface nodes by x-coordinate for spatial locality
    raw_x = x[:, :, 0]
    sort_idx = raw_x.argsort(dim=1)
    pred_sorted = torch.gather(surf_pred_p_masked, 1, sort_idx)
    mask_sorted = torch.gather(surf_mask, 1, sort_idx)
    # Rolling variance over windows of 16 nodes
    window = 16
    B_s, N_s = pred_sorted.shape
    if N_s >= window:
        pred_unfold = pred_sorted.unfold(1, window, 1)  # [B, N-w+1, w]
        mask_unfold = mask_sorted.unsqueeze(-1).float().unfold(1, window, 1).squeeze(-2)
        local_mean = (pred_unfold * mask_unfold).sum(-1) / mask_unfold.sum(-1).clamp(min=1)
        local_var = ((pred_unfold - local_mean.unsqueeze(-1))**2 * mask_unfold).sum(-1) / mask_unfold.sum(-1).clamp(min=1)
        # Pad to full length
        pad_left = window // 2
        pad_right = N_s - local_var.shape[1] - pad_left
        local_var_full = F.pad(local_var, (pad_left, pad_right), value=0.0)
        # Unsort back to original order
        unsort_idx = sort_idx.argsort(dim=1)
        local_var_orig = torch.gather(local_var_full, 1, unsort_idx)
        # Weight: 1.0 base + 0.5 * normalized_variance
        var_norm = local_var_orig / (local_var_orig.max(dim=1, keepdim=True).values + 1e-8)
        hardness_weight = (1.0 + 0.5 * var_norm).unsqueeze(-1)  # [B, N, 1]
        # Apply to surface pressure error
        surf_pres_weighted = abs_err[:, :, 2:3] * hardness_weight
        surf_per_sample = (surf_pres_weighted * surf_mask.unsqueeze(-1)).sum(dim=(1, 2)) / surf_mask.sum(dim=1).clamp(min=1).float()
```

This replaces the hard-mining block's `surf_per_sample` computation when epoch >= 15.

Run with `--wandb_group flow-gradient-loss`

## Baseline
val_loss=0.8477 | in_dist=17.74 | ood_cond=13.77 | ood_re=27.52 | tandem=37.72

---

## Results
_To be filled by student_

### CODE REVIEW

**Comment by tcapelle** [comment]
Closing — 'haku' is not a valid student. Orphaned PR.
